lib/Fhp/PaginateableAction.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Fhp;
+
+use Fhp\Protocol\BPD;
+use Fhp\Protocol\Message;
+use Fhp\Protocol\UnexpectedResponseException;
+use Fhp\Protocol\UPD;
+use Fhp\Segment\HIRMS\Rueckmeldungscode;
+use Fhp\Segment\Paginateable;
+
+abstract class PaginateableAction extends BaseAction
+{
+    /**
+     * @var Stores the request created by BaseAction::getNextRequest to be reused in cases where the bank response
+     * can be split over multiple pages e.g. request/response pairs. This avoids the need for {@link BPD} to be
+     * available for paginated requests.
+     */
+    private $requestSegments;
+
+    /**
+     * If set, the last response from the server regarding this action indicated that there are more results to be
+     * fetched using this pagination token. This is called "Aufsetzpunkt" in the specification.
+     * @var string|null
+     */
+    private $paginationToken;
+
+    /** {@inheritdoc} */
+    public function serialize(): string
+    {
+        if (!$this->needsTan()) {
+            throw new \RuntimeException('Cannot serialize this action, because it is not waiting for a TAN.');
+        }
+        return serialize([$this->requestSegmentNumbers, $this->tanRequest, $this->needTanForSegment, $this->paginationToken, $this->requestSegments]);
+    }
+
+    /** {@inheritdoc} */
+    public function unserialize($serialized)
+    {
+        list($this->requestSegmentNumbers, $this->tanRequest, $this->needTanForSegment, $this->paginationToken, $this->requestSegments) = unserialize($serialized);
+    }
+
+    /**
+     * @return bool True if the response has not been read completely yet, i.e. additional requests to the server are
+     *     necessary to continue reading the requested data.
+     */
+    public function hasMorePages(): bool
+    {
+        return !$this->isDone && $this->paginationToken !== null;
+    }
+
+    /** {@inheritdoc} */
+    public function processResponse(Message $response)
+    {
+        $pagination = $response->findRueckmeldung(Rueckmeldungscode::PAGINATION);
+        if ($pagination === null) {
+            $this->paginationToken = null;
+            parent::processResponse($response);
+        } else {
+            if (count($pagination->rueckmeldungsparameter) !== 1) {
+                throw new UnexpectedResponseException("Unexpected pagination request: $pagination");
+            }
+            $this->paginationToken = $pagination->rueckmeldungsparameter[0];
+        }
+    }
+
+    /** {@inheritdoc} */
+    public function getNextRequest(?BPD $bpd, ?UPD $upd)
+    {
+        if ($this->requestSegments === null) {
+            $this->requestSegments = parent::getNextRequest($bpd, $upd);
+        } else {
+            if ($this->paginationToken !== null) {
+                foreach ($this->requestSegments as $segment) {
+                    if ($segment instanceof Paginateable) {
+                        $segment->setPaginationToken($this->paginationToken);
+                    }
+                }
```

Comment:
Ich würde schon hier `$this->paginationToken = null;` setzen und dafür dann in `processResponse()` nicht.